File: examples/android/RunAnywhereAI/app/src/main/java/com/runanywhere/runanywhereai/presentation/voice/TranscriptionViewModel.kt
package com.runanywhere.runanywhereai.presentation.voice

import android.app.Application
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.runanywhere.sdk.events.EventBus
import com.runanywhere.sdk.public.RunAnywhere
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.serialization.Serializable
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.UUID

/**
 * Transcript segment with speaker diarization support
 * Matches iOS TranscriptSegment
 */
@Serializable
data class TranscriptSegment(
    val id: String = UUID.randomUUID().toString(),
    val text: String,
    val timestamp: Long = System.currentTimeMillis(),
    val isFinal: Boolean,
    val speakerId: String? = null,
    val speakerName: String? = null
)

/**
 * Speaker information for diarization
 * Matches iOS SpeakerInfo
 */
@Serializable
data class SpeakerInfo(
    val id: String,
    val name: String? = null,
    val color: Int? = null
)

// Note: ModelLoadState is defined in VoiceAssistantViewModel.kt

/**
 * TranscriptionViewModel - Voice transcription with speaker diarization
 * Matches iOS TranscriptionViewModel.swift structure
 */
class TranscriptionViewModel(application: Application) : AndroidViewModel(application) {

    // MARK: - Published Properties (matching iOS)

    private val _transcriptionText = MutableStateFlow("")
    val transcriptionText: StateFlow<String> = _transcriptionText.asStateFlow()

    private val _isTranscribing = MutableStateFlow(false)
    val isTranscribing: StateFlow<Boolean> = _isTranscribing.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    private val _isInitialized = MutableStateFlow(false)
    val isInitialized: StateFlow<Boolean> = _isInitialized.asStateFlow()

    private val _currentStatus = MutableStateFlow("Initializing...")
    val currentStatus: StateFlow<String> = _currentStatus.asStateFlow()

    private val _whisperModel = MutableStateFlow("Whisper Base")
    val whisperModel: StateFlow<String> = _whisperModel.asStateFlow()

    private val _partialTranscript = MutableStateFlow("")
    val partialTranscript: StateFlow<String> = _partialTranscript.asStateFlow()

    private val _finalTranscripts = MutableStateFlow<List<TranscriptSegment>>(emptyList())
    val finalTranscripts: StateFlow<List<TranscriptSegment>> = _finalTranscripts.asStateFlow()

    private val _detectedSpeakers = MutableStateFlow<List<SpeakerInfo>>(emptyList())
    val detectedSpeakers: StateFlow<List<SpeakerInfo>> = _detectedSpeakers.asStateFlow()

    private val _currentSpeaker = MutableStateFlow<SpeakerInfo?>(null)
    val currentSpeaker: StateFlow<SpeakerInfo?> = _currentSpeaker.asStateFlow()

    private val _enableSpeakerDiarization = MutableStateFlow(true)
    val enableSpeakerDiarization: StateFlow<Boolean> = _enableSpeakerDiarization.asStateFlow()

    // MARK: - Model Loading State (from SDK lifecycle tracker)

    private val _sttModelState = MutableStateFlow(ModelLoadState.NOT_LOADED)
    val sttModelState: StateFlow<ModelLoadState> = _sttModelState.asStateFlow()

    val isModelLoaded: Boolean get() = _sttModelState.value.isLoaded

    // MARK: - Private Properties

    private val whisperModelName: String = "whisper-base"

    // MARK: - Initialization

    init {
        initialize()
    }

    private fun initialize() {
        viewModelScope.launch {
            try {
                // TODO: Request microphone permission
                // Matches iOS: await AudioCapture.requestMicrophonePermission()

                // Initialize SDK
                RunAnywhere.initialize(
                    apiKey = "demo-api-key",
                    baseURL = "https://api.runanywhere.ai",
                    environment = com.runanywhere.sdk.data.models.SDKEnvironment.DEVELOPMENT
                )

                // Subscribe to model lifecycle changes
                subscribeToModelLifecycle()

                // Update whisper model display name
                updateWhisperModelName()

                // Start observing SDK events
                observeSDKEvents()

                _currentStatus.value = "Ready (FluidAudio)"
                _isInitialized.value = true

            } catch (e: Exception) {
                _errorMessage.value = "Failed to initialize: ${e.message}"
                _currentStatus.value = "Error"
            }
        }
    }

    /**
     * Subscribe to SDK's model lifecycle tracker for STT model state updates
     * Matches iOS subscribeToModelLifecycle()
     */
    private fun subscribeToModelLifecycle() {
        viewModelScope.launch {
            // TODO: Observe changes to STT model via the SDK's lifecycle tracker
            // Matches iOS: ModelLifecycleTracker.shared.$modelsByModality
            //     .receive(on: DispatchQueue.main)
            //     .sink { modelsByModality in ... }

            // For now, simulate initial state
            _sttModelState.value = ModelLoadState.NOT_LOADED
        }
    }

    private fun updateWhisperModelName() {
        _whisperModel.value = when (whisperModelName) {
            "whisper-base" -> "Whisper Base"
            "whisper-small" -> "Whisper Small"
            "whisper-medium" -> "Whisper Medium"
            "whisper-large" -> "Whisper Large"
            "whisper-large-v3" -> "Whisper Large v3"
            else -> whisperModelName.replace("-", " ")
                .replaceFirstChar { it.uppercase() }
        }
    }

    // MARK: - Transcription Control

    /**
     * Start real-time transcription
     * Matches iOS startTranscription()
     */
    fun startTranscription() {
        if (_isTranscribing.value) {
            return
        }

        viewModelScope.launch {
            try {
                // TODO: Create modular pipeline config matching iOS
                // val config = ModularPipelineConfig.transcriptionWithVAD(
                //     sttModel = whisperModelName,
                //     vadThreshold = 0.01f
                // )

                // TODO: Create pipeline with optional diarization
                // if (_enableSpeakerDiarization.value) {
                //     voicePipeline = FluidAudioIntegration.createVoicePipelineWithDiarization(config)
                // } else {
                //     voicePipeline = RunAnywhere.createVoicePipeline(config)
                // }

                // TODO: Initialize components
                // for (event in pipeline.initializeComponents()) { ... }

                // TODO: Start audio capture and process through pipeline
                // val audioStream = audioCapture.startContinuousCapture()
                // pipelineTask = launch { pipeline.process(audioStream) }

                _isTranscribing.value = true
                _currentStatus.value = "Listening..."
                _errorMessage.value = null
                _partialTranscript.value = ""

            } catch (e: Exception) {
                _errorMessage.value = "Failed to start transcription: ${e.message}"
                _currentStatus.value = "Error"
            }
        }
    }

    /**
     * Stop transcription
     * Matches iOS stopTranscription()
     */
    fun stopTranscription() {
        if (!_isTranscribing.value) {
            return
        }

        viewModelScope.launch {
            try {
                // TODO: Cancel pipeline task
                // pipelineTask?.cancel()

                // TODO: Stop audio capture
                // audioCapture.stopContinuousCapture()

                _isTranscribing.value = false
                _currentStatus.value = "Ready to transcribe"

                // Add final partial transcript if exists
                if (_partialTranscript.value.isNotEmpty()) {
                    val segment = TranscriptSegment(
                        text = _partialTranscript.value,
                        isFinal = true,
                        speakerId = _currentSpeaker.value?.id,
                        speakerName = _currentSpeaker.value?.name
                    )
                    _finalTranscripts.value = _finalTranscripts.value + segment
                    _partialTranscript.value = ""
                }

            } catch (e: Exception) {
                _errorMessage.value = "Failed to stop transcription: ${e.message}"
            }
        }
    }

    /**
     * Clear all transcripts
     * Matches iOS clearTranscripts()
     */
    fun clearTranscripts() {
        _finalTranscripts.value = emptyList()
        _partialTranscript.value = ""
        _transcriptionText.value = ""
    }

    /**
     * Export transcripts as text
     * Matches iOS exportTranscripts()
     */
    fun exportTranscripts(): String {
        val dateFormatter = SimpleDateFormat("MM/dd/yy h:mm:ss a", Locale.getDefault())
        val now = Date()

        val builder = StringBuilder()
        builder.appendLine("Transcription Export")
        builder.appendLine("Date: ${dateFormatter.format(now)}")
        builder.appendLine("Model: ${_whisperModel.value}")
        builder.appendLine("---")
        builder.appendLine()

        for (segment in _finalTranscripts.value) {
            val timestamp = Date(segment.timestamp)
            builder.appendLine("[${dateFormatter.format(timestamp)}]")
            if (segment.speakerName != null) {
                builder.appendLine("Speaker: ${segment.speakerName}")
            }
            builder.appendLine(segment.text)
            builder.appendLine()
        }

        if (_partialTranscript.value.isNotEmpty()) {
            builder.appendLine("[Current]")
            builder.appendLine(_partialTranscript.value)
        }

        return builder.toString()
    }

    /**
     * Copy all transcripts to clipboard
     * Matches iOS copyToClipboard()
     */
    fun copyToClipboard() {
        val fullText = _finalTranscripts.value.joinToString(" ") { it.text }
        val textToCopy = if (fullText.isEmpty()) _partialTranscript.value else fullText

        val clipboard = getApplication<Application>()
            .getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        val clip = ClipData.newPlainText("Transcription", textToCopy)
        clipboard.setPrimaryClip(clip)
    }

    /**
     * Toggle speaker diarization
     */
    fun toggleSpeakerDiarization() {
        _enableSpeakerDiarization.value = !_enableSpeakerDiarization.value
    }

    /**
     * Clear error message
     */
    fun clearError() {
        _errorMessage.value = null
    }

    // MARK: - SDK Event Observation

    private fun observeSDKEvents() {
        viewModelScope.launch {
            // Observe voice events from EventBus
            EventBus.voiceEvents.collect { event ->
                handleVoiceEvent(event)
            }
        }
    }

    private fun handleVoiceEvent(event: com.runanywhere.sdk.events.SDKVoiceEvent) {
        when (event) {
            is com.runanywhere.sdk.events.SDKVoiceEvent.TranscriptionStarted -> {
                _isTranscribing.value = true
                _currentStatus.value = "Listening..."
                _partialTranscript.value = ""
            }

            is com.runanywhere.sdk.events.SDKVoiceEvent.TranscriptionPartial -> {
                _partialTranscript.value = event.text
            }

            is com.runanywhere.sdk.events.SDKVoiceEvent.TranscriptionFinal -> {
                if (event.text.isNotBlank()) {
                    val segment = TranscriptSegment(
                        text = event.text,
                        isFinal = true,
                        speakerId = _currentSpeaker.value?.id,
                        speakerName = _currentSpeaker.value?.name
                    )
                    _finalTranscripts.value = _finalTranscripts.value + segment
                    _partialTranscript.value = ""
                    _transcriptionText.value = _finalTranscripts.value.joinToString(" ") { it.text }
                }
            }

            is com.runanywhere.sdk.events.SDKVoiceEvent.PipelineError -> {
                _errorMessage.value = event.error.message ?: "Unknown error"
                _currentStatus.value = "Error"
                _isTranscribing.value = false
            }

            else -> {
                // Handle other events as needed
            }
        }
    }
}

File: examples/ios/RunAnywhereAI/RunAnywhereAI/Features/Voice/TranscriptionViewModel.swift
import Foundation
import RunAnywhereSDK
import AVFoundation
import Combine
import os
#if os(iOS) || targetEnvironment(macCatalyst)
import UIKit
#elseif os(macOS)
import AppKit
#endif

@MainActor
class TranscriptionViewModel: ObservableObject {
    private let logger = Logger(subsystem: "com.runanywhere.RunAnywhereAI", category: "TranscriptionViewModel")
    private let sdk = RunAnywhereSDK.shared
    private let audioCapture = AudioCapture()

    // MARK: - Published Properties
    @Published var transcriptionText: String = ""
    @Published var isTranscribing: Bool = false
    @Published var errorMessage: String?
    @Published var isInitialized = false
    @Published var currentStatus = "Initializing..."
    @Published var whisperModel: String = "Whisper Base"
    @Published var partialTranscript: String = ""
    @Published var finalTranscripts: [TranscriptSegment] = []
    @Published var detectedSpeakers: [SpeakerInfo] = []
    @Published var currentSpeaker: SpeakerInfo?
    @Published var enableSpeakerDiarization: Bool = true

    // MARK: - Transcription State
    private var voicePipeline: VoicePipelineManager?
    private let whisperModelName: String = "whisper-base"
    private var audioStreamContinuation: AsyncStream<VoiceAudioChunk>.Continuation?
    private var pipelineTask: Task<Void, Never>?

    // MARK: - Transcript Segment Model
    struct TranscriptSegment: Identifiable {
        let id = UUID()
        let text: String
        let timestamp: Date
        let isFinal: Bool
        let speaker: SpeakerInfo?

        init(text: String, timestamp: Date, isFinal: Bool, speaker: SpeakerInfo? = nil) {
            self.text = text
            self.timestamp = timestamp
            self.isFinal = isFinal
            self.speaker = speaker
        }
    }

    // MARK: - Initialization

    func initialize() async {
        logger.info("Initializing TranscriptionViewModel...")

        // Request microphone permission
        logger.info("Requesting microphone permission...")
        let hasPermission = await AudioCapture.requestMicrophonePermission()
        logger.info("Microphone permission: \(hasPermission)")

        guard hasPermission else {
            currentStatus = "Microphone permission denied"
            errorMessage = "Please enable microphone access in Settings"
            logger.error("Microphone permission denied")
            return
        }

        // Set the Whisper model display name
        updateWhisperModelName()

        // FluidAudio is always available
        logger.info("✅ FluidAudioDiarization module is available")
        currentStatus = "Ready (FluidAudio)"

        logger.info("Transcription service initialized")
        isInitialized = true
    }

    private func updateWhisperModelName() {
        switch whisperModelName {
        case "whisper-base":
            whisperModel = "Whisper Base"
        case "whisper-small":
            whisperModel = "Whisper Small"
        case "whisper-medium":
            whisperModel = "Whisper Medium"
        case "whisper-large":
            whisperModel = "Whisper Large"
        case "whisper-large-v3":
            whisperModel = "Whisper Large v3"
        default:
            whisperModel = whisperModelName.replacingOccurrences(of: "-", with: " ").capitalized
        }
        logger.info("Using Whisper model: \(self.whisperModel)")
    }

    // MARK: - Transcription Control

    /// Start real-time transcription
    func startTranscription() async {
        guard !isTranscribing else {
            logger.warning("Transcription already in progress")
            return
        }

        logger.info("Starting transcription with modular pipeline...")

        // Create a simple transcription-only pipeline config
        // Lower VAD threshold from 0.02 to 0.01 for better voice detection
        let config = ModularPipelineConfig.transcriptionWithVAD(
            sttModel: whisperModelName,
            vadThreshold: 0.01  // Lowered for more sensitive voice detection
        )

        // Create the pipeline with FluidAudio diarization if enabled
        if enableSpeakerDiarization {
            logger.info("Using FluidAudioDiarization for speaker detection")
            voicePipeline = await FluidAudioIntegration.createVoicePipelineWithDiarization(
                sdk: sdk,
                config: config
            )
            voicePipeline?.enableSpeakerDiarization(true)
            voicePipeline?.enableContinuousMode(true)
        } else {
            // Create standard pipeline
            voicePipeline = sdk.createVoicePipeline(config: config)
            voicePipeline?.delegate = self

            // Enable speaker diarization with default implementation
            if enableSpeakerDiarization {
                voicePipeline?.enableSpeakerDiarization(true)
                voicePipeline?.enableContinuousMode(true)
                logger.info("Enabled speaker diarization with default implementation")
            }
        }

        voicePipeline?.delegate = self

        // Initialize components first (VAD and STT only for transcription)
        guard let pipeline = voicePipeline else {
            errorMessage = "Failed to create pipeline"
            currentStatus = "Error"
            logger.error("Failed to create transcription pipeline")
            return
        }

        // Initialize all components
        currentStatus = "Initializing components..."
        do {
            for try await event in pipeline.initializeComponents() {
                if case .componentInitialized(let name) = event {
                    logger.info("Initialized: \(name)")
                } else if case .componentInitializationFailed(let name, let error) = event {
                    logger.error("Failed to initialize \(name): \(error)")
                    throw error
                }
            }
        } catch {
            errorMessage = "Component initialization failed: \(error.localizedDescription)"
            currentStatus = "Error"
            logger.error("Component initialization failed: \(error)")
            return
        }

        // Start audio capture and process through pipeline
        let audioStream = audioCapture.startContinuousCapture()

        isTranscribing = true
        currentStatus = "Listening..."
        errorMessage = nil
        partialTranscript = ""

        // Process audio through pipeline
        pipelineTask = Task {
            do {
                for try await event in pipeline.process(audioStream: audioStream) {
                    await handlePipelineEvent(event)
                }
            } catch {
                await MainActor.run {
                    self.errorMessage = "Pipeline error: \(error.localizedDescription)"
                    self.currentStatus = "Error"
                    self.isTranscribing = false
                }
            }
        }

        logger.info("Transcription pipeline started successfully")
    }

    /// Stop transcription
    func stopTranscription() async {
        guard isTranscribing else {
            logger.warning("No transcription in progress")
            return
        }

        logger.info("Stopping transcription...")

        // Cancel pipeline task
        pipelineTask?.cancel()
        pipelineTask = nil

        // Stop audio capture
        audioCapture.stopContinuousCapture()

        voicePipeline = nil

        isTranscribing = false
        currentStatus = "Ready to transcribe"

        // Add final partial transcript if exists
        if !partialTranscript.isEmpty {
            finalTranscripts.append(TranscriptSegment(
                text: partialTranscript,
                timestamp: Date(),
                isFinal: true,
                speaker: currentSpeaker
            ))
            partialTranscript = ""
        }

        logger.info("Transcription stopped")
    }

    /// Clear all transcripts
    func clearTranscripts() {
        finalTranscripts.removeAll()
        partialTranscript = ""
        transcriptionText = ""
        logger.info("Transcripts cleared")
    }

    /// Export transcripts as text
    func exportTranscripts() -> String {
        let dateFormatter = DateFormatter()
        dateFormatter.dateStyle = .short
        dateFormatter.timeStyle = .medium

        var exportText = "Transcription Export\n"
        exportText += "Date: \(dateFormatter.string(from: Date()))\n"
        exportText += "Model: \(whisperModel)\n"
        exportText += "---\n\n"

        for segment in finalTranscripts {
            exportText += "[\(dateFormatter.string(from: segment.timestamp))]\n"
            exportText += "\(segment.text)\n\n"
        }

        if !partialTranscript.isEmpty {
            exportText += "[Current]\n"
            exportText += "\(partialTranscript)\n"
        }

        return exportText
    }

    /// Copy all transcripts to clipboard
    func copyToClipboard() {
        let fullText = finalTranscripts.map { $0.text }.joined(separator: " ")
        let textToCopy = fullText.isEmpty ? partialTranscript : fullText

        #if os(iOS) || targetEnvironment(macCatalyst)
        UIPasteboard.general.string = textToCopy
        #elseif os(macOS)
        let pasteboard = NSPasteboard.general
        pasteboard.clearContents()
        pasteboard.setString(textToCopy, forType: .string)
        #endif

        logger.info("Copied \(textToCopy.count) characters to clipboard")
    }

    // MARK: - Pipeline Event Handling

    private func handlePipelineEvent(_ event: ModularPipelineEvent) async {
        await MainActor.run {
            switch event {
            case .vadSpeechStart:
                logger.info("Speech detected")
                currentStatus = "Listening..."

            case .vadSpeechEnd:
                logger.info("Speech ended")

            case .sttPartialTranscript(let text):
                partialTranscript = text
                logger.info("Partial transcript: '\(text)'")

            case .sttFinalTranscript(let text):
                if !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                    finalTranscripts.append(TranscriptSegment(
                        text: text,
                        timestamp: Date(),
                        isFinal: true,
                        speaker: nil
                    ))
                    partialTranscript = ""
                    transcriptionText = finalTranscripts.map { $0.text }.joined(separator: " ")
                }
                logger.info("Final transcript: '\(text)'")

            case .sttFinalTranscriptWithSpeaker(let text, let speaker):
                if !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                    finalTranscripts.append(TranscriptSegment(
                        text: text,
                        timestamp: Date(),
                        isFinal: true,
                        speaker: speaker
                    ))
                    partialTranscript = ""
                    transcriptionText = finalTranscripts.map { $0.text }.joined(separator: " ")
                    currentSpeaker = speaker

                    // Update detected speakers list
                    if !detectedSpeakers.contains(where: { $0.id == speaker.id }) {
                        detectedSpeakers.append(speaker)
                    }
                }
                logger.info("Final transcript from \(speaker.name ?? speaker.id): '\(text)'")

            case .sttPartialTranscriptWithSpeaker(let text, let speaker):
                partialTranscript = text
                currentSpeaker = speaker
                logger.info("Partial transcript from \(speaker.name ?? speaker.id): '\(text)'")

            case .sttNewSpeakerDetected(let speaker):
                if !detectedSpeakers.contains(where: { $0.id == speaker.id }) {
                    detectedSpeakers.append(speaker)
                }
                logger.info("New speaker detected: \(speaker.name ?? speaker.id)")

            case .sttSpeakerChanged(let from, let to):
                currentSpeaker = to
                logger.info("Speaker changed from \(from?.name ?? from?.id ?? "unknown") to \(to.name ?? to.id)")

            case .pipelineError(let error):
                errorMessage = error.localizedDescription
                currentStatus = "Error"
                logger.error("Pipeline error: \(error)")

            default:
                // Ignore other events not relevant to transcription
                break
            }
        }
    }
}

// MARK: - VoicePipelineManagerDelegate

extension TranscriptionViewModel: @preconcurrency VoicePipelineManagerDelegate {
    nonisolated func pipeline(_ pipeline: VoicePipelineManager, didReceiveEvent event: ModularPipelineEvent) {
        Task { @MainActor in
            await handlePipelineEvent(event)
        }
    }

    nonisolated func pipeline(_ pipeline: VoicePipelineManager, didEncounterError error: Error) {
        Task { @MainActor in
            errorMessage = error.localizedDescription
            isTranscribing = false
            currentStatus = "Error"
        }
        logger.error("Pipeline error: \(error)")
    }
}
